src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,215 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Diagnostics;
+using System.Text.Json.Serialization;
 
 namespace System.Text.Json
 {
     public static partial class JsonSerializer
     {
-        private static void HandleMetadataPropertyValue(ref Utf8JsonReader reader, ref ReadStack state)
+        internal static bool ResolveMetadata<T>(
+            this JsonConverter converter,
+            ref Utf8JsonReader reader,
+            ref ReadStack state,
+            out T value)
         {
-            Debug.Assert(state.Current.JsonClassInfo!.Options.ReferenceHandling.ShouldReadPreservedReferences());
-
-            if (reader.TokenType != JsonTokenType.String)
+            if (state.Current.ObjectState < StackFrameObjectState.MetadataPropertyName)
             {
-                ThrowHelper.ThrowJsonException_MetadataValueWasNotString(reader.TokenType);
+                // Read the first metadata property name.
+                if (!reader.Read())
+                {
+                    value = default!;
+                    return false;
+                }
+
+                if (reader.TokenType != JsonTokenType.PropertyName)
+                {
+                    // An enumerable needs metadata since it starts with StartObject.
+                    if (converter.ClassType == ClassType.Enumerable)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(converter.TypeToConvert);
+                    }
+
+                    // The reader should have detected other invalid cases.
+                    Debug.Assert(reader.TokenType == JsonTokenType.EndObject);
+
+                    value = default!;
+                    return true;
+                }
+
+                ReadOnlySpan<byte> propertyName = GetSpan(ref reader);
+                MetadataPropertyName metadata = GetMetadataPropertyName(propertyName);
+                state.Current.JsonPropertyName = propertyName.ToArray();
```

Comment:
Move this to the branches other than `NoMetadata`. Otherwise, we set `JsonPropertyName` as the first property and then we never set it again which forces the JSON path to always show the first property.

```cs
[Fact]
public static void DictionaryKeyPath()
{
    string json = @"{""Key1"":{""$id"":""1""},""Key2"":{""Name"":""A"",""$ref"":""1""}}";
    JsonException ex = Assert.Throws<JsonException>(() => JsonSerializer.Deserialize<Dictionary<string, Employee>>(json, s_deserializerOptionsPreserve));
    
    Assert.Equal("$.Key2.$ref", ex.Path); // ex.Path is $.Key1.$ref here.
```